nft_ingester/src/main.rs

```diff
@@ -1,59 +1,41 @@
-mod bubblegum_indexer;
-mod crud_indexer;
-mod gummyroll_indexer;
-mod utils;
-
 use {
-    crate::{
-        bubblegum_indexer::handle_bubblegum_instruction,
-        crud_indexer::handle_gummyroll_crud_instruction,
-        gummyroll_indexer::handle_gummyroll_instruction,
-        utils::{filter_events_from_logs, order_instructions},
+    futures_util::TryFutureExt,
+    messenger::{ACCOUNT_STREAM, TRANSACTION_STREAM},
+    nft_ingester::parsers::{
+        BubblegumHandler, GummyRollHandler, InstructionBundle, ProgramHandler,
+        ProgramHandlerManager,
     },
-    flatbuffers::{ForwardsUOffset, Vector},
-    lazy_static::lazy_static,
-    messenger::{ACCOUNT_STREAM, BLOCK_STREAM, SLOT_STREAM, TRANSACTION_STREAM},
+    nft_ingester::utils::{order_instructions, parse_logs},
     plerkle::async_redis_messenger::AsyncRedisMessenger,
+    plerkle_serialization::account_info_generated::account_info::root_as_account_info,
     plerkle_serialization::transaction_info_generated::transaction_info::root_as_transaction_info,
-    regex::Regex,
     solana_sdk::pubkey::Pubkey,
-    sqlx::{self, postgres::PgPoolOptions, Pool, Postgres},
+    sqlx::{self, postgres::PgPoolOptions},
+    std::sync::Arc,
 };
 
-mod program_ids {
-    #![allow(missing_docs)]
-
-    use solana_sdk::pubkeys;
-    pubkeys!(
-        token_metadata,
-        "metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s"
-    );
-    pubkeys!(
-        gummyroll_crud,
-        "Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS"
-    );
-    pubkeys!(bubblegum, "BGUMzZr2wWfD2yzrXFEWTK2HbdYhqQCP2EZoPEkZBD6o");
-    pubkeys!(gummyroll, "GRoLLMza82AiYN7W9S9KCCtCyyPRAQP2ifBy4v4D5RMD");
-    pubkeys!(token, "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb");
-    pubkeys!(a_token, "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
-}
-
-#[tokio::main]
-async fn main() {
-    // Setup Postgres.
+async fn setup_manager<'c>(mut manager: ProgramHandlerManager<'c>) -> ProgramHandlerManager<'c> {
+    // TODO setup figment gor db configuration
     let pool = PgPoolOptions::new()
         .max_connections(5)
         .connect("postgres://solana:solana@db/solana")
         .await
         .unwrap();
```

Comment:
So each time a manager is setup there is a new Pool with 5 connections created?  I thought the idea of the sqlx Pool struct was to create it once and then just send copies of the handle around to different tasks.

From the sqlx core docs:
```
/// `Pool` is `Send`, `Sync` and `Clone`. It is intended to be created once at the start of your
/// application/daemon/web server/etc. and then shared with all tasks throughout the process'
/// lifetime. How best to accomplish this depends on your program architecture.
```
Seems like you could still just pass in a copy to service_transaction_stream:
```
tasks.push(service_transaction_stream(pool.clone()).await);
```
and then you could just pass it into the constructor:
```
let mut manager = ProgramHandlerManager::new(pool);
```
Seems simpler than having to use arc for this use case, but I'm not adamant about this, more just curious of the need for arc for this particular object.

Reply:
According to the [docs for pool](https://docs.rs/sqlx/latest/sqlx/struct.Pool.html) and seeing what ppl do online, in the simplest terms we should be able to do something like:
```
#[tokio::main]
async fn main() {
    let pool = PgPoolOptions::new()
        .max_connections(5)
        .connect("postgres://solana:solana@db/solana")
        .await
        .unwrap();

   let mut tasks = vec![];

    for i in 0..10 {
        let pool = pool.clone();

        tasks.push(tokio::spawn(async move {
            service_something(i, &pool).await;
        }));
    }
}
```
So each task just gets a copy of the connection pool.  It doesn't seem like we need arc or a mutex.  When I read [this section of the docs](https://docs.rs/sqlx/latest/sqlx/struct.Pool.html#why-use-a-pool) it seems like just using the Pool object as-is is meant to solve those issues of reference counting and mutual exclusion on any given connection in the pool.

Caveats:
1.  There used to be [a bug](https://github.com/launchbadge/sqlx/issues/622) with some of this behavior but looks like it was fixed (although potential repro 5 days ago, but not sure if usage is correct, it looks slightly different).
2. You might be seeing the future state more than me with registering more parsers and separating account handlers from transaction handlers, etc. so maybe I'm not seeing the architectural reason to have a new pool per `ProgramHandlerManager`.

Reply:
looks like this was resolved in most recent change